pandas/core/reshape/reshape.py

```diff
@@ -604,12 +611,28 @@ def _convert_level_number(level_num, columns):
             chunk.columns = level_vals.take(chunk.columns.labels[-1])
             value_slice = chunk.reindex(columns=level_vals_used).values
         else:
-            if frame._is_mixed_type:
+            if (frame._is_homogeneous_type and
+                    is_extension_array_dtype(frame.dtypes.iloc[0])):
+                dtype = this.loc[:, this.columns[loc]].dtypes.iloc[0]
+                subset = this.loc[:, this.columns[loc]]
+
+                value_slice = dtype.construct_array_type()._concat_same_type(
+                    [x._values for _, x in subset.iteritems()]
+                )
+                N, K = this.shape
```

Comment:
maybe make a little helper function out of these 3 lines (so you can use it above as well)